🐛 BUG: `wrangler dev --experimental-local` slows down over time and leaks memory

### What version of `Wrangler` are you using?

2.6.2

### What operating system are you using?

Mac

### Describe the Bug

If I test my simple worker with `wrangler dev --experimental-local` and monitor the browser devtools network tab I see that after the initial (cold) requests then next requests return in 2ms.

If I keep on refreshing the browser, I see that time slowly starts to creep up, 4ms, 6ms, 8ms, and if I keep on refreshing the browser tab a lot, it slows down to 50-70ms and keeps on slowing down further with additional http requests.

At the same time I observe that the memory usage increases from 17MB to 34MB, 50MB, 75MB, all the way to over 1GB.

Hopefully this will help in some small way, but I've had issues with the slowdown and the bus 10 errors as well.

Started at 3.0.1, and now on 3.1.0 on MAC

Both gave me a `Bus error: 10` message along with massive slow-downs out the gate.

I used this command, `softwareupdate --install "Command Line Tools for Xcode-14.3"`, and that _seemed_ to get rid of the bus 10 error. The speed was still abysmal, taking a couple of minutes to load in vs the fractions of a second that it takes on prod.

Command used:

```
wrangler dev --log-level debug --ip 0.0.0.0 --port 3000 --env development
```

I did, however, get a different error than the bus 10 error after the CLI XCode update:

```
workerd/jsg/util.c++:278: info: exception = kj/async-io-unix.c++:778: disconnected: ::writev(fd, iov.begin(), iov.size()): Broken pipe; iovTotal = 1108; iov.size() = 1
 stack: 10210f62f 1021100d3 10124d3cc 1020e7df8 1020ec88c 1020e37a8 1020f2c54 1009f95e8 1020f16d0 100df1130
 workerd/io/worker.c++:1703: info: uncaught exception; source = Uncaught (in promise); exception = Error: Network connection lost.
 workerd/io/worker.c++:1703: info: uncaught exception; source = Uncaught (in response); exception = Error: Network connection lost.
 workerd/jsg/util.c++:278: info: exception = kj/async-io-unix.c++:778: disconnected: ::writev(fd, iov.begin(), iov.size()): Broken pipe; iovTotal = 1108; iov.size() = 1
 stack: 10210f62f 1021100d3 10124d3cc 1020e7df8 1020e802f 1020d349c 1020e7df8 1020e802f 1020d349c 1020e7df8 1020e802f 1020d349c 1020f9314 1020f16d0 100f77494 1009f95e8 100e9e7c0 100dd8868 100eaf900 100d40c08 100df01c8 100b074b4
 workerd/io/worker.c++:1703: info: uncaught exception; source = Uncaught (in promise); exception = Error: Network connection lost.
 workerd/io/worker.c++:1703: info: uncaught exception; source = Uncaught (async); exception = Error: Network connection lost.
 ✘ [ERROR] Uncaught (in promise) Error: Network connection lost.
 
 
 ✘ [ERROR] Uncaught (async) Error: Network connection lost.
```

For additional context, the application runs as expected on the miniflare 2.X versions. 

Command used with miniflare:

```
miniflare --live-reload --debug --port 3000 --env .env --modules <script path>
```

I hope this isn't just adding static

--- a/.changeset/empty-avocados-promise.md
+++ b/.changeset/empty-avocados-promise.md
@@ -1,0 +1,8 @@
+---
+"wrangler": patch
+---
+
+fix: register middleware once for module workers
+
+Ensure middleware is only registered on the first request when using module workers.
+This should prevent stack overflows and slowdowns when making large number of requests to `wrangler dev` with infrequent reloads.

--- a/packages/wrangler/src/__tests__/middleware.test.ts
+++ b/packages/wrangler/src/__tests__/middleware.test.ts
@@ -1084,6 +1084,7 @@ describe("middleware", () => {
 			  }
 			  return env;
 			}
+			var registeredMiddleware = false;
 			var facade2 = {
 			  ...middleware_insertion_facade_default.tail && {
 			    tail: maskHandlerEnv(middleware_insertion_facade_default.tail)
@@ -1103,8 +1104,11 @@ describe("middleware", () => {
 			  fetch(request, rawEnv, ctx) {
 			    const env = getMaskedEnv2(rawEnv);
 			    if (middleware_insertion_facade_default.middleware && middleware_insertion_facade_default.middleware.length > 0) {
-			      for (const middleware of middleware_insertion_facade_default.middleware) {
-			        __facade_register__(middleware);
+			      if (!registeredMiddleware) {
+			        registeredMiddleware = true;
+			        for (const middleware of middleware_insertion_facade_default.middleware) {
+			          __facade_register__(middleware);
+			        }
 			      }
 			      const __facade_modules_dispatch__ = function(type, init) {
 			        if (type === \\"scheduled\\" && middleware_insertion_facade_default.scheduled !== void 0) {

--- a/packages/wrangler/templates/middleware/loader-modules.ts
+++ b/packages/wrangler/templates/middleware/loader-modules.ts
@@ -65,6 +65,8 @@ type MissingExportHandlers = Omit<
 	"tail" | "trace" | "scheduled" | "queue" | "test" | "fetch"
 >;
 
+let registeredMiddleware = false;
+
 const facade: ExportedHandler<unknown> & MissingExportHandlers = {
 	...(worker.tail && {
 		tail: maskHandlerEnv(worker.tail),
@@ -86,8 +88,13 @@ const facade: ExportedHandler<unknown> & MissingExportHandlers = {
 		const env = getMaskedEnv(rawEnv);
 		// Get the chain of middleware from the worker object
 		if (worker.middleware && worker.middleware.length > 0) {
-			for (const middleware of worker.middleware) {
-				__facade_register__(middleware);
+			// Make sure we only register middleware once:
+			// https://github.com/cloudflare/workers-sdk/issues/2386#issuecomment-1614715911
+			if (!registeredMiddleware) {
+				registeredMiddleware = true;
+				for (const middleware of worker.middleware) {
+					__facade_register__(middleware);
+				}
 			}
 
 			const __facade_modules_dispatch__: Dispatcher = function (type, init) {

